### PR TITLE
8311591: Add SystemModulesPlugin test case that splits module descriptors with new local variables defined by DedupSetBuilder 

### PR DESCRIPTION
**Pull request by Siedlerchr**
Add new test case with sample modules that contains some requires/exports/uses/provides.

We are just unsure if and how we should add some last step of verificaiton with the extracted and decompiled class.

Follow up task from https://github.com/openjdk/jdk/pull/14408

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311591](https://bugs.openjdk.org/browse/JDK-8311591): Add SystemModulesPlugin test case that splits module descriptors with new local variables defined by DedupSetBuilder (**Bug** - P3)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [2bcdd2fc](https://git.openjdk.org/jdk/pull/15234/files/2bcdd2fc645d360992f48b0639891afd246e7ca3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15234/head:pull/15234` \
`$ git checkout pull/15234`

Update a local copy of the PR: \
`$ git checkout pull/15234` \
`$ git pull https://git.openjdk.org/jdk.git pull/15234/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15234`

View PR using the GUI difftool: \
`$ git pr show -t 15234`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15234.diff">https://git.openjdk.org/jdk/pull/15234.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15234#issuecomment-1673979294)

### CODE REVIEW

**Comment by koppor** [comment]
We accidently pushed the full internal history - and "immediatly" squashed into one commit.

**Review by mlchung — COMMENTED** [review]
Thanks for adding this case.   The steps to extract from jimage and inspect the generated bytecode of `SystemModules$all` is for us to manually verify if the test does the work, i.e. each `subX` method adds more local variables due to deduplication.   It's not needed in the test itself.

Since the batch size is 1, I would suggest that `p4.Main` can also load `jdk.internal.module.SystemModules\$all` and verify the expected numbers of `subX` methods (one per each module).   To find all modules in the image, you can simply do `ModuleFinder.ofSystem().findAll()`.

BTW, `dedup/src/*` files should also have the copyright header.

**Comment by Siedlerchr** [comment]
> Since the batch size is 1, I would suggest that `p4.Main` can also load `jdk.internal.module.SystemModules\$all` and verify the expected numbers of `subX` methods (one per each module). To find all modules in the image, you can simply do `ModuleFinder.ofSystem().findAll()`.

Thanks for the suggestion. We added now a check in the main class for the number of generated sub methods. We put a fixed number, as we did not get the relation between the number of "require modules" statements and the total number of modules.

File: test/jdk/tools/jlink/JLinkDedupTestBatchSizeOne.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                "-m", "m4/p4.Main");
+        processBuilder.inheritIO();
+        processBuilder.directory(binDir.toFile());
+        Process process = processBuilder.start();
+        int exitCode = process.waitFor();
+        if (exitCode != 0)
+            throw new AssertionError("JLinkDedupTest100Modules failed to launch");
+
+        extractJImage(image);
+        decompileWitJavap(image);
```
Comment (mlchung) [inline test/jdk/tools/jlink/JLinkDedupTestBatchSizeOne.java:116]: No needed.   These 2 lines along with the methods can be deleted.

File: test/jdk/tools/jlink/JLinkDedupTestBatchSizeOne.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    public static void main(String[] args) throws Throwable {
+        compileAll();
+        Path image = Paths.get("bug8311591");
+
+        JImageGenerator.getJLinkTask()
+                .modulePath(MODULE_PATH)
+                .output(image.resolve("out-jlink-dedup"))
+                .addMods("m1")
+                .addMods("m2")
+                .addMods("m2")
```
Comment (mlchung) [inline test/jdk/tools/jlink/JLinkDedupTestBatchSizeOne.java:94]: `m2` is added twice.  The duplicated line should be removed.

File: test/jdk/tools/jlink/JLinkDedupTestBatchSizeOne.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import tests.Result;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/*
+ * @test
+ * @summary Make sure that modules can be linked using jlink and deduplication works correctly when creating sub methods
```
Comment (mlchung) [inline test/jdk/tools/jlink/JLinkDedupTestBatchSizeOne.java:38]: please wrap this long line.

File: test/jdk/tools/jlink/dedup/src/m1/p1/AInterface.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package p1;
```
Comment (mlchung) [inline test/jdk/tools/jlink/dedup/src/m1/p1/AInterface.java:1]: missing copyright header.

File: test/jdk/tools/jlink/dedup/src/m4/p4/Main.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        for (var service : services) {
+            System.out.println("Service name  " + service.getServiceName());
+            System.out.println("Service string " + service.getString());
+        }
+        var moduleClass = Class.forName("jdk.internal.module.SystemModules$all");
+        long subMethodCount = Arrays.stream(moduleClass.getDeclaredMethods()).filter(method -> method.getName().startsWith("sub")).count();
+
+        if (subMethodCount != MODULE_SUB_METHOD_COUNT) {
+            throw new AssertionError("Difference in generated sub module methods count! Expected: " + MODULE_SUB_METHOD_COUNT + " but was " + subMethodCount);
+        }
```
Comment (mlchung) [inline test/jdk/tools/jlink/dedup/src/m4/p4/Main.java:47]:
```suggestion
        // one subX method per each module is generated as the image is linked with
        // --system-modules=batchSize=1
        var moduleCount = ModuleFinder.ofSystem().findAll().stream().count();
        if (subMethodCount != moduleCount) {
            throw new AssertionError("Difference in generated sub module methods count! Expected: " +
                    moduleCount + " but was " + subMethodCount);
        }
```

File: test/jdk/tools/jlink/dedup/src/m4/p4/Main.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    private final static int MODULE_SUB_METHOD_COUNT = 9;
+
+    public static void main(String[] args) throws Exception {
+        List<ServiceInterface> services = getServices();
+        for (var service : services) {
+            System.out.println("Service name  " + service.getServiceName());
+            System.out.println("Service string " + service.getString());
+        }
+        var moduleClass = Class.forName("jdk.internal.module.SystemModules$all");
+        long subMethodCount = Arrays.stream(moduleClass.getDeclaredMethods()).filter(method -> method.getName().startsWith("sub")).count();
```
Comment (mlchung) [inline test/jdk/tools/jlink/dedup/src/m4/p4/Main.java:43]: nit: wrap long line

```suggestion
        long subMethodCount = Arrays.stream(moduleClass.getDeclaredMethods())
                                    .filter(method -> method.getName().startsWith("sub"))
                                    .count();
```

**Review by mlchung — COMMENTED** [review]
Looks good.   Since the image is created with batchSize = 1, it means that one `subX` method is created for each module linked in the image.  So the test can count the number of modules in the image and the number of `subX` methods should match the module count.  I include the suggested code to check that.

**Comment by Siedlerchr** [comment]
Thanks a lot for your explanation and suggestions! Now we understood this :)

**Review by mlchung — COMMENTED** [review]
JLinkDedupTestBatchSizeOne.java line 115-144 can be deleted; otherwise, looks good.

**Review by mlchung — APPROVED** [review]

File: test/jdk/tools/jlink/dedup/src/m4/p4/Main.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import p3.ServiceInterface;
+
+import java.lang.module.ModuleFinder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceLoader;
+
+public class Main {
+    private final static int MODULE_SUB_METHOD_COUNT = 9;
```
[the remark this answers is not in the document]
Comment (Siedlerchr) [reply test/jdk/tools/jlink/dedup/src/m4/p4/Main.java]: Ah yeah overlooked that!

**Comment by Siedlerchr** [comment]
@mlchung Thanks again for your help. What would be needed to get the at https://github.com/openjdk/jdk/commit/ec7da91bd83803b7d91a4de3a01caf0ba256c037 back ported to JDK21? I guess for the initial release it's too late but maybe for one of the jdk21u releases? That would be really a huge step forward for us at JabRef. (Currently we rely on a custom build)

**Comment by mlchung** [comment]
I created the backport PRs for [JDK-8240567](https://bugs.openjdk.org/browse/JDK-8240567) and [JDK-8311591](https://bugs.openjdk.org/browse/JDK-8311591).

https://github.com/openjdk/jdk21u/pull/65
https://github.com/openjdk/jdk21u/pull/66

**Pull request closed without merging**